presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.classloader_interface;
+
+import org.apache.spark.Partition;
+import org.apache.spark.SparkContext;
+import org.apache.spark.TaskContext;
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.rdd.ZippedPartitionsBaseRDD;
+import org.apache.spark.rdd.ZippedPartitionsPartition;
+import scala.Tuple2;
+import scala.collection.Seq;
+import scala.reflect.ClassTag;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyIterator;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+import static scala.collection.JavaConversions.asJavaIterator;
+import static scala.collection.JavaConversions.asScalaBuffer;
+import static scala.collection.JavaConversions.asScalaIterator;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+public class PrestoSparkTaskRdd
```

Comment:
Let's add some comment for this class. Feel free to adapt/modify the following:

```
/**
 * PrestoSparkTaskRdd represents execution of Presto stage, it contains:
 *  - a list of shuffleInputRdds, each of the corresponding to a child stage.
 *  - an optional taskSourceRdd, which represents *ALL* the table scan inputs in this stage.
 *
 *  Table scan presents when join bucketed table with unbucketed table, for example:
 *        Join
 *        /  \
 *    Scan  Remote
 *  In this case, bucket to Spark partition mapping has to be consistent with the Spark shuffle partition.
 *
 *  shuffleInputRdds can also be empty when the stage partitioning is SINGLE_DISTRIBUTION.
 *  TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION.
 *
 *  The broadcast input is encapsulated in taskProcessor.
 */
```

Let me know if the TODO suggestion (TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION) makes sense. Since I personally found it difficult to reason when both `shuffleInputRdds` and `taskSourceRdd` can be empty.

Reply:
> (TODO: Consider have dedicated RDD implementation for SINGLE_DISTRIBUTION) 

I'm not particularly convinced that a separate RDD specifically for single distribution makes a lot of sense, especially if sometimes single distributed fragment may still have inputs. It might make things even more confusing.

Reply:
@arhimondr : Yeah, let's make how to handle `SINGLE_DISTRIBUTION` as a future work :)